In [1]:
from collections import defaultdict, OrderedDict
from csv import reader
from itertools import chain, combinations
from optparse import OptionParser
import pandas as pd

# if not installed yet: pip install mlxtend
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import fpgrowth, association_rules

###  How to find frequent itemsets using FPGrowth?

In [2]:
dataset = [['Corn', 'Light Cream', 'Chicken', 'Beef', 'Wine', 'Ice Cream'],
           ['Dill', 'Onion', 'Carrot', 'Beef', 'Wine', 'Ice Cream'],
           ['Milk', 'Wine', 'Beef', 'Ice Cream'],
           ['Light Cream', 'Chicken', 'Corn', 'Kidney Beans', 'Yogurt', 'Wine'],
           ['Corn', 'Onion', 'Light Cream', 'Kidney Beans', 'Chicken', 'Yogurt']]

te = TransactionEncoder()
te_ary = te.fit(dataset).transform(dataset)
df = pd.DataFrame(te_ary, columns=te.columns_)

frequent_itemsets = fpgrowth(df, min_support=0.6, use_colnames=True)
frequent_itemsets

support                      itemsets
0       0.8                        (Wine)
1       0.6                 (Light Cream)
2       0.6                   (Ice Cream)
3       0.6                        (Corn)
4       0.6                     (Chicken)
5       0.6                        (Beef)
6       0.6             (Ice Cream, Wine)
7       0.6           (Light Cream, Corn)
8       0.6               (Chicken, Corn)
9       0.6        (Light Cream, Chicken)
10      0.6  (Light Cream, Chicken, Corn)
11      0.6             (Ice Cream, Beef)
12      0.6                  (Wine, Beef)
13      0.6       (Ice Cream, Wine, Beef)

## Part 1(a): 
### What association rules can be found in given dataset with the minimum support 60% and the minimum confidence 70% ?
If we are only interested in rules with the confidence level above the 70 percent threshold (min_threshold=0.7), we can find the set of rules with the following specifications accordingly using *generate_rules()* function.
1. Specify your metric of interest 
2. Threshold

In [3]:
association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)

antecedents             consequents  antecedent support  \
0              (Ice Cream)                  (Wine)                 0.6   
1                   (Wine)             (Ice Cream)                 0.8   
2            (Light Cream)                  (Corn)                 0.6   
3                   (Corn)           (Light Cream)                 0.6   
4                (Chicken)                  (Corn)                 0.6   
5                   (Corn)               (Chicken)                 0.6   
6            (Light Cream)               (Chicken)                 0.6   
7                (Chicken)           (Light Cream)                 0.6   
8   (Light Cream, Chicken)                  (Corn)                 0.6   
9      (Light Cream, Corn)               (Chicken)                 0.6   
10         (Chicken, Corn)           (Light Cream)                 0.6   
11           (Light Cream)         (Chicken, Corn)                 0.6   
12               (Chicken)     (Light Cream, Corn)                 0.6   
13                  (Corn)  (Light Cream, Chicken)                 0.6   
14             (Ice Cream)                  (Beef)                 0.6   
15                  (Beef)             (Ice Cream)                 0.6   
16                  (Wine)                  (Beef)                 0.8   
17                  (Beef)                  (Wine)                 0.6   
18       (Ice Cream, Wine)                  (Beef)                 0.6   
19       (Ice Cream, Beef)                  (Wine)                 0.6   
20            (Wine, Beef)             (Ice Cream)                 0.6   
21             (Ice Cream)            (Wine, Beef)                 0.6   
22                  (Wine)       (Ice Cream, Beef)                 0.8   
23                  (Beef)       (Ice Cream, Wine)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.8      0.6        1.00  1.250000      0.12         inf  
1                  0.6      0.6        0.75  1.250000      0.12         1.6  
2                  0.6      0.6        1.00  1.666667      0.24         inf  
3                  0.6      0.6        1.00  1.666667      0.24         inf  
4                  0.6      0.6        1.00  1.666667      0.24         inf  
5                  0.6      0.6        1.00  1.666667      0.24         inf  
6                  0.6      0.6        1.00  1.666667      0.24         inf  
7                  0.6      0.6        1.00  1.666667      0.24         inf  
8                  0.6      0.6        1.00  1.666667      0.24         inf  
9                  0.6      0.6        1.00  1.666667      0.24         inf  
10                 0.6      0.6        1.00  1.666667      0.24         inf  
11                 0.6      0.6        1.00  1.666667      0.24         inf  
12                 0.6      0.6        1.00  1.666667      0.24         inf  
13                 0.6      0.6        1.00  1.666667      0.24         inf  
14                 0.6      0.6        1.00  1.666667      0.24         inf  
15                 0.6      0.6        1.00  1.666667      0.24         inf  
16                 0.6      0.6        0.75  1.250000      0.12         1.6  
17                 0.8      0.6        1.00  1.250000      0.12         inf  
18                 0.6      0.6        1.00  1.666667      0.24         inf  
19                 0.8      0.6        1.00  1.250000      0.12         inf  
20                 0.6      0.6        1.00  1.666667      0.24         inf  
21                 0.6      0.6        1.00  1.666667      0.24         inf  
22                 0.6      0.6        0.75  1.250000      0.12         1.6  
23                 0.6      0.6        1.00  1.666667      0.24         inf

## Part 1(b): 
### What association rules can be found in given dataset with the minimum support 60% and the minimum lift value 1.2 ?

In [4]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1.2)
rules

antecedents             consequents  antecedent support  \
0              (Ice Cream)                  (Wine)                 0.6   
1                   (Wine)             (Ice Cream)                 0.8   
2            (Light Cream)                  (Corn)                 0.6   
3                   (Corn)           (Light Cream)                 0.6   
4                (Chicken)                  (Corn)                 0.6   
5                   (Corn)               (Chicken)                 0.6   
6            (Light Cream)               (Chicken)                 0.6   
7                (Chicken)           (Light Cream)                 0.6   
8   (Light Cream, Chicken)                  (Corn)                 0.6   
9      (Light Cream, Corn)               (Chicken)                 0.6   
10         (Chicken, Corn)           (Light Cream)                 0.6   
11           (Light Cream)         (Chicken, Corn)                 0.6   
12               (Chicken)     (Light Cream, Corn)                 0.6   
13                  (Corn)  (Light Cream, Chicken)                 0.6   
14             (Ice Cream)                  (Beef)                 0.6   
15                  (Beef)             (Ice Cream)                 0.6   
16                  (Wine)                  (Beef)                 0.8   
17                  (Beef)                  (Wine)                 0.6   
18       (Ice Cream, Wine)                  (Beef)                 0.6   
19       (Ice Cream, Beef)                  (Wine)                 0.6   
20            (Wine, Beef)             (Ice Cream)                 0.6   
21             (Ice Cream)            (Wine, Beef)                 0.6   
22                  (Wine)       (Ice Cream, Beef)                 0.8   
23                  (Beef)       (Ice Cream, Wine)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.8      0.6        1.00  1.250000      0.12         inf  
1                  0.6      0.6        0.75  1.250000      0.12         1.6  
2                  0.6      0.6        1.00  1.666667      0.24         inf  
3                  0.6      0.6        1.00  1.666667      0.24         inf  
4                  0.6      0.6        1.00  1.666667      0.24         inf  
5                  0.6      0.6        1.00  1.666667      0.24         inf  
6                  0.6      0.6        1.00  1.666667      0.24         inf  
7                  0.6      0.6        1.00  1.666667      0.24         inf  
8                  0.6      0.6        1.00  1.666667      0.24         inf  
9                  0.6      0.6        1.00  1.666667      0.24         inf  
10                 0.6      0.6        1.00  1.666667      0.24         inf  
11                 0.6      0.6        1.00  1.666667      0.24         inf  
12                 0.6      0.6        1.00  1.666667      0.24         inf  
13                 0.6      0.6        1.00  1.666667      0.24         inf  
14                 0.6      0.6        1.00  1.666667      0.24         inf  
15                 0.6      0.6        1.00  1.666667      0.24         inf  
16                 0.6      0.6        0.75  1.250000      0.12         1.6  
17                 0.8      0.6        1.00  1.250000      0.12         inf  
18                 0.6      0.6        1.00  1.666667      0.24         inf  
19                 0.8      0.6        1.00  1.250000      0.12         inf  
20                 0.6      0.6        1.00  1.666667      0.24         inf  
21                 0.6      0.6        1.00  1.666667      0.24         inf  
22                 0.6      0.6        0.75  1.250000      0.12         1.6  
23                 0.6      0.6        1.00  1.666667      0.24         inf

## Part 1(c):
### Examine the differences from the two sets of associations rules above. Select one rule that has low confidence and high lift value. Why is this rule interesting?

### Answer: 
1(a) and 1(b) have identical rules. All rules with 1.2 lift have more than 70% confidence and vice versa.    
If I were to select one interesting rule - Rule 1 from 1(b) above has a confidence value of 0.75 and a lift value of 1.25.
This rule states that (Wine -> Icecream).  

Confidence of 0.75 means that when 'Wine' is part of a transaction, our rule (Wine,Icecream) is observed 75% of the time. In other words, we are only 75% certain that 'Icecream' will also be in the transaction when we see 'Wine' in the transaction.  

Lift of 1.25 means that 'Wine' and 'Icecream' are positive correlated - P(Icecream|Wine) = 3/4 = 0.75 , while P(Icecream) = 3/5 = 0.6, therefore we calculate a lift value of 0.75/0.6 which is 1.25, which shows that the combination of Icecream/Wine is more frequent than a sampling for just Icecream randomly.  

This rule is of high interest because P(Icecream,Wine) = 0.6 while P(Icecream) x P(Wine) = 0.6x0.8, which makes its interest 0.6/(0.6x0.8) = 1.25. This further confirms that the two items are positively correlated.

## Part 1(d):
### Describing results: A summary of results (number of rules, general description), and a selection of rules that you deem is interesting based on “interestingness" measure(s) of your choice. Describe your reasoning behind any choices you made.

### Answer:
A total of 18 rules were identified based on my new custom interestingness threshold of 60% support and min 1.5 lift. Below is a selection of rules I have deemed as 'Interesting'.   

I chose these new thresholds because a higher lift value signifies a higher positive correlation between the items. The default output from earlier parts did not feel like they were significant enough so I decided to raise the lift threshold to 1.5.  

This new threshold will only identify rules of high lift, meaning consequent items are at minimum 1.5 times more likely to appear in rows that contain the antecdent, when comared to consequent items appearing in the entire dataset. ( P(con|ant) 1.5 times more likely than P(con) ). This essentially suggests that the correlation between these pairs are less likely to be caused by chance while randomly sampling, and more likely an actual underlying pattern in the dataset.


In [19]:
rules2 = association_rules(frequent_itemsets, metric="lift", min_threshold=1.5)
rules2

antecedents             consequents  antecedent support  \
0            (Light Cream)                  (Corn)                 0.6   
1                   (Corn)           (Light Cream)                 0.6   
2                (Chicken)                  (Corn)                 0.6   
3                   (Corn)               (Chicken)                 0.6   
4            (Light Cream)               (Chicken)                 0.6   
5                (Chicken)           (Light Cream)                 0.6   
6   (Light Cream, Chicken)                  (Corn)                 0.6   
7      (Light Cream, Corn)               (Chicken)                 0.6   
8          (Chicken, Corn)           (Light Cream)                 0.6   
9            (Light Cream)         (Chicken, Corn)                 0.6   
10               (Chicken)     (Light Cream, Corn)                 0.6   
11                  (Corn)  (Light Cream, Chicken)                 0.6   
12             (Ice Cream)                  (Beef)                 0.6   
13                  (Beef)             (Ice Cream)                 0.6   
14       (Ice Cream, Wine)                  (Beef)                 0.6   
15            (Wine, Beef)             (Ice Cream)                 0.6   
16             (Ice Cream)            (Wine, Beef)                 0.6   
17                  (Beef)       (Ice Cream, Wine)                 0.6   

    consequent support  support  confidence      lift  leverage  conviction  
0                  0.6      0.6         1.0  1.666667      0.24         inf  
1                  0.6      0.6         1.0  1.666667      0.24         inf  
2                  0.6      0.6         1.0  1.666667      0.24         inf  
3                  0.6      0.6         1.0  1.666667      0.24         inf  
4                  0.6      0.6         1.0  1.666667      0.24         inf  
5                  0.6      0.6         1.0  1.666667      0.24         inf  
6                  0.6      0.6         1.0  1.666667      0.24         inf  
7                  0.6      0.6         1.0  1.666667      0.24         inf  
8                  0.6      0.6         1.0  1.666667      0.24         inf  
9                  0.6      0.6         1.0  1.666667      0.24         inf  
10                 0.6      0.6         1.0  1.666667      0.24         inf  
11                 0.6      0.6         1.0  1.666667      0.24         inf  
12                 0.6      0.6         1.0  1.666667      0.24         inf  
13                 0.6      0.6         1.0  1.666667      0.24         inf  
14                 0.6      0.6         1.0  1.666667      0.24         inf  
15                 0.6      0.6         1.0  1.666667      0.24         inf  
16                 0.6      0.6         1.0  1.666667      0.24         inf  
17                 0.6      0.6         1.0  1.666667      0.24         inf

# Part 2:

In [17]:
class Node:
    def __init__(self, itemName, frequency, parentNode):
        self.itemName = itemName
        self.count = frequency
        self.parent = parentNode
        self.children = {}
        self.next = None

    def increment(self, frequency):
        self.count += frequency

    def display(self, ind=1):
        print('  ' * ind, self.itemName, ' ', self.count)
        for child in list(self.children.values()):
            child.display(ind+1)


def constructTree(itemSetList, frequency, minSup):
    headerTable = defaultdict(int)
    # Counting frequency and create header table
    for idx, itemSet in enumerate(itemSetList):
        for item in itemSet:
            headerTable[item] += frequency[idx]

    # Deleting items below minSup
    headerTable = dict((item, sup) for item, sup in headerTable.items() if sup >= minSup)
    if(len(headerTable) == 0):
        return None, None

    # HeaderTable column [Item: [frequency, headNode]]
    for item in headerTable:
        headerTable[item] = [headerTable[item], None]

    # Init Null head node
    fpTree = Node('Null', 1, None)
    # Update FP tree for each cleaned and sorted itemSet
    for idx, itemSet in enumerate(itemSetList):
        itemSet = [item for item in itemSet if item in headerTable]
        # 2(b) Code modification
        itemSet.sort(reverse = True)
        itemSet.sort(key=lambda item: headerTable[item][0], reverse=True)
        # Traverse from root to leaf, update tree with given item
        currentNode = fpTree
        for item in itemSet:
            currentNode = updateTree(item, currentNode, headerTable, frequency[idx])

    return fpTree, headerTable

def updateHeaderTable(item, targetNode, headerTable):
    if(headerTable[item][1] == None):
        headerTable[item][1] = targetNode
    else:
        currentNode = headerTable[item][1]
        # Traverse to the last node then link it to the target
        while currentNode.next != None:
            currentNode = currentNode.next
        currentNode.next = targetNode

def updateTree(item, treeNode, headerTable, frequency):
    if item in treeNode.children:
        # If the item already exists, increment the count
        treeNode.children[item].increment(frequency)
    else:
        # Create a new branch
        newItemNode = Node(item, frequency, treeNode)
        treeNode.children[item] = newItemNode
        # Link the new branch to header table
        updateHeaderTable(item, newItemNode, headerTable)

    return treeNode.children[item]

def ascendFPtree(node, prefixPath):
    if node.parent != None:
        prefixPath.append(node.itemName)
        ascendFPtree(node.parent, prefixPath)

def findPrefixPath(basePat, headerTable):
    # First node in linked list
    treeNode = headerTable[basePat][1] 
    condPats = []
    frequency = []
    while treeNode != None:
        prefixPath = []
        # From leaf node all the way to root
        ascendFPtree(treeNode, prefixPath)  
        if len(prefixPath) > 1:
            # Storing the prefix path and it's corresponding count
            condPats.append(prefixPath[1:])
            frequency.append(treeNode.count)

        # Go to next node
        treeNode = treeNode.next  
    return condPats, frequency

def mineTree(headerTable, minSup, preFix, freqItemList):
    # Sort the items with frequency and create a list
    sortedItemList = [item[0] for item in sorted(list(headerTable.items()), key=lambda p:p[1][0])] 
    # Start with the lowest frequency
    for item in sortedItemList:  
        # Pattern growth is achieved by the concatenation of suffix pattern with frequent patterns generated from conditional FP-tree
        newFreqSet = preFix.copy()
        newFreqSet.add(item)
        freqItemList.append(newFreqSet)
        # Find all prefix path, constrcut conditional pattern base
        conditionalPattBase, frequency = findPrefixPath(item, headerTable) 
        # Construct conditonal FP Tree with conditional pattern base
        conditionalTree, newHeaderTable = constructTree(conditionalPattBase, frequency, minSup) 
        if newHeaderTable != None:
            # Mining recursively on the tree
            mineTree(newHeaderTable, minSup,
                       newFreqSet, freqItemList)

def powerset(s):
    return chain.from_iterable(combinations(s, r) for r in range(1, len(s)))

def getSupport(testSet, itemSetList):
    count = 0
    for itemSet in itemSetList:
        if(set(testSet).issubset(itemSet)):
            count += 1
    return count

def associationRule(freqItemSet, itemSetList, minConf):
    rules = []
    for itemSet in freqItemSet:
        subsets = powerset(itemSet)
        itemSetSup = getSupport(itemSet, itemSetList)
        for s in subsets:
            confidence = float(itemSetSup / getSupport(s, itemSetList))
            if(confidence > minConf):
                rules.append([set(s), set(itemSet.difference(s)), confidence])
    return rules

def getFrequencyFromList(itemSetList):
    frequency = [1 for i in range(len(itemSetList))]
    return frequency

In [9]:
def fpgrowth(itemSetList, minSupRatio, minConf):
    frequency = getFrequencyFromList(itemSetList)
    
    minSup = len(itemSetList) * minSupRatio
#     print(minSup)
#     print(itemSetList)
#     print(frequency)
    fpTree, headerTable = constructTree(itemSetList, frequency, minSup)
    
    if(fpTree == None):
        print('No frequent item set')
    else:
        freqItems = []
        mineTree(headerTable, minSup, set(), freqItems)
        rules = associationRule(freqItems, itemSetList, minConf)
        return freqItems, rules

## Part 2(a):
### What association rules can be found with the code provided in Part 2 with the minimum support 60% and the minimum confidence 70%? Comparing to the output from question 1(a), what do you think the cause is that leads to different outcome?

### Answer:
Only 14 rules are mined with the default part 2 code.  
My initial thought as of this point after examining the code leads me to believe that the way the item set is being sorted in the provided code is causing this behaviour.

In [9]:
freqItemSet, rules = fpgrowth(dataset, minSupRatio=0.6, minConf=0.7)
freqItemSet, rules

([{'Corn'},
  {'Light Cream'},
  {'Chicken'},
  {'Chicken', 'Light Cream'},
  {'Beef'},
  {'Beef', 'Wine'},
  {'Ice Cream'},
  {'Beef', 'Ice Cream'},
  {'Ice Cream', 'Wine'},
  {'Beef', 'Ice Cream', 'Wine'},
  {'Wine'}],
 [[{'Light Cream'}, {'Chicken'}, 1.0],
  [{'Chicken'}, {'Light Cream'}, 1.0],
  [{'Wine'}, {'Beef'}, 0.75],
  [{'Beef'}, {'Wine'}, 1.0],
  [{'Ice Cream'}, {'Beef'}, 1.0],
  [{'Beef'}, {'Ice Cream'}, 1.0],
  [{'Ice Cream'}, {'Wine'}, 1.0],
  [{'Wine'}, {'Ice Cream'}, 0.75],
  [{'Beef'}, {'Ice Cream', 'Wine'}, 1.0],
  [{'Ice Cream'}, {'Beef', 'Wine'}, 1.0],
  [{'Wine'}, {'Beef', 'Ice Cream'}, 0.75],
  [{'Beef', 'Ice Cream'}, {'Wine'}, 1.0],
  [{'Beef', 'Wine'}, {'Ice Cream'}, 1.0],
  [{'Ice Cream', 'Wine'}, {'Beef'}, 1.0]])

## Part 2(b):
### Fix the problem in the Part 2 code and re-run your program on the given dataset to obtain the same results as in Part 1(a). What did you change?

### Answer:
The itemSet was not being sorted correctly in the constructTree function. Once I found the correct place to be sorting the itemSet and corrected this behaviour, part 2 started producing the same rules as part 1, albeit in different ordering. (Same 24 rules, different ordering)

**itemSet.sort(reverse = True)** was added in the constructTree function

In [18]:
freqItemSet, rules = fpgrowth(dataset, minSupRatio=0.6, minConf=0.7)
freqItemSet, rules

([{'Beef'},
  {'Beef', 'Ice Cream'},
  {'Beef', 'Ice Cream', 'Wine'},
  {'Beef', 'Wine'},
  {'Chicken'},
  {'Chicken', 'Corn'},
  {'Chicken', 'Corn', 'Light Cream'},
  {'Chicken', 'Light Cream'},
  {'Corn'},
  {'Corn', 'Light Cream'},
  {'Ice Cream'},
  {'Ice Cream', 'Wine'},
  {'Light Cream'},
  {'Wine'}],
 [[{'Ice Cream'}, {'Beef'}, 1.0],
  [{'Beef'}, {'Ice Cream'}, 1.0],
  [{'Ice Cream'}, {'Beef', 'Wine'}, 1.0],
  [{'Wine'}, {'Beef', 'Ice Cream'}, 0.75],
  [{'Beef'}, {'Ice Cream', 'Wine'}, 1.0],
  [{'Ice Cream', 'Wine'}, {'Beef'}, 1.0],
  [{'Beef', 'Ice Cream'}, {'Wine'}, 1.0],
  [{'Beef', 'Wine'}, {'Ice Cream'}, 1.0],
  [{'Wine'}, {'Beef'}, 0.75],
  [{'Beef'}, {'Wine'}, 1.0],
  [{'Chicken'}, {'Corn'}, 1.0],
  [{'Corn'}, {'Chicken'}, 1.0],
  [{'Light Cream'}, {'Chicken', 'Corn'}, 1.0],
  [{'Chicken'}, {'Corn', 'Light Cream'}, 1.0],
  [{'Corn'}, {'Chicken', 'Light Cream'}, 1.0],
  [{'Chicken', 'Light Cream'}, {'Corn'}, 1.0],
  [{'Corn', 'Light Cream'}, {'Chicken'}, 1.0],
  [{'Chicken

## Part 2(c):
### Explain in your own words how the anti-monotonicity property of itemsets is maintained to reduce the search space in generating frequent itemsets.

### Answer:

According to the anti-monotone property: the support of a superset can never exceed the support of its subset.
For example, if item [D] is found to only appear in 1 transaction in the whole database (support of 1), and the minimum support threshold is set at 2, then any supersets of D [AD, BD, ABD, ABCD] can never exceed a support level of 1 (otherwise [D] would have a higher support).   

This property can be applied during rule generation as once we know the set [D] is infrequent, then any superset that contains [D] will also be infrequent so we can skip those transactions instead of wasting time scanning and calculating support redundantly. This process is also called pruning.

## Part 2(d):
### Explain in your own words how the pattern growth is achieved in generating frequent itemsets.

### Answer:
Firstly the entire database is scanned to count the frequency of each unique item and how often they appear(support). 
After the transactions are fully scanned we sort this list of frequent 1-items in (typically) descending support. If the support for the 1-item is below the minimum threshold then the item is discarded.   

Next, a null root node is created and the entire database is scanned again, each transaction will have its frequent items sorted in descending support according to the table we created before. A branch is created atop the null root node in this sorted order.  

This process is repeated for every transaction in the database. If the sorted ordering of a transaction already exists, then each node in the branch would simply increase by a count of 1. Else new nodes are created for patterns that are completely different or partially different to existing branch.   

Finally, nodes with the same name are linked together to create the finished FP-Tree.  
The FP-Tree is traversed in ascending count recursively to determine the frequent itemsets.


## Part 2(e):
### Explain in your own words how many data scan is required for FPgrowth compared to Apriori.

### Answer:  
For FP-Growth: Only 2 scans are required.   
The first scan through the database counts the frequency each item appears and sorts the results, then the second scan of the database creates the root node, header table, and FP tree (If a frequent item is present in that transaction).  

For Apriori: In the WORST case, 2^k scans are required (where k is the number of unique items in the database. 2^k-2 if we exclude the rules that have an empty LHS or RHS, but these do not impact time complexity much in the bigger picture so I left them out of the discussion)  
Firstly, the database is scanned for frequent 1-item sets.   
After all the frequent 1-items are discovered we then scan for all frequent 2-itemsets that start with every 1-itemset (immediate superset).   
After all the frequent 2-itemsets are found we then scan for frequent 3-itemsets that begin with every 2-itemset and so on until no more association rules can be found.   
Essentially the antimonotone properties as described previously are applied at every iteration to prune out rules that we know are not frequent.    
The worst case is observed when either every item in the database meets the minimum support threshold, or if the minimum support is set to 0.  

## Part 2(f):
### Explain in your own words how time complexity and memory consumption is reduced compared to Apriori.

### Answer:
As mentioned in the previous question, the complexity for FP-Growth is linear(2*rows) while the complexity of Apriori can reach 2^k in the worst case where every item in the database meets minimum support threshold.

Also, Apriori uses Arrays as its storage structure, while FP-Growth uses Trees as its storage structure. Trees are more efficient than arrays when inserting new records therefore it is more memory efficient. Trees also require significantly less memory than Arrays.   

Apriori also generates a list of all candidates and then prunes the useless ones, while FP-Growth only generates the known frequent candidates. Therefore Apriori has a higher memory demand when compared to FP-Growth as it has to store all the possible candidate sets at some point.   

Next, Apriori takes a Breadth First Search approach while FP-Growth takes a divide-and-conquer approach. As a result of this FP-Growth time requirements grow linearly as more items are considered, while Apriori grows exponentially.    

The divide-and-conquer approach can also be done in parallel which further reduces its time complexity.   
In summary: FP-Growth is more efficient in both memory and time when compared to Apriori.

# Extra Marks
### Please provide a different item ordering (i.e., not descending order of frequency) in constructing FP-Tree and draw the corresponding FP-Tree ofthe same datasetas above. Do you believe that you have the most compressed (or compact) FP-Tree representation in this particular case? 

### Answer:
I would order the items in alphabetical order. I believe this is the most compact possible FP-Tree as there is no way to sort the itemset in a more definitive order.

<img src="image.png">
 